controlplane/internal/controller/clusterdeployment_controller.go

```diff
@@ -295,38 +303,87 @@ type InstallConfigOverride struct {
-// the OpenshiftAssistedControlPlane has additional capabilities set.
-// For MNO (multi-node openshift), this is set to the default list for baremetal platform,
-// but the OpenshiftAssistedControlPlane can specify additional capabilities to be appended to this list.
-func setACICapabilities(oacp *controlplanev1alpha2.OpenshiftAssistedControlPlane, aci *hiveext.AgentClusterInstall) error {
-	isBaremetalPlatform := aci.Spec.PlatformType == hiveext.BareMetalPlatformType
-	if isCapabilitiesEmpty(oacp.Spec.Config.Capabilities) && !isBaremetalPlatform {
-		return nil
+// isBaremetalPlatform checks if the AgentClusterInstall is configured for a baremetal platform.
+// Returns true if the platform type is BareMetalPlatformType, false otherwise.
+func isBaremetalPlatform(aci *hiveext.AgentClusterInstall) bool {
+	return aci.Spec.PlatformType == hiveext.BareMetalPlatformType
+}
+
+// getInstallConfigOverride merges install config override from annotations with capabilities-based overrides.
+// It returns the final merged install config override as a JSON string, or nil if no overrides are needed.
+// The function combines user-provided install config overrides from annotations with automatically
+// generated capabilities configuration based on the cluster platform type.
+func getInstallConfigOverride(oacp *controlplanev1alpha2.OpenshiftAssistedControlPlane, aci *hiveext.AgentClusterInstall) (*string, error) {
+	var installConfigOverrideStr string
+	cfg, hasInstallConfigOverride := oacp.Annotations[controlplanev1alpha2.InstallConfigOverrideAnnotation]
+	if hasInstallConfigOverride {
+		installConfigOverrideStr = cfg
+	}
+	capabilitiesCfgOverride, err := getInstallConfigOverrideForCapabilities(oacp, aci)
+	if err != nil {
+		return nil, err
+	}
+	// if no override and no capabilities, no need to merge install config override
+	if installConfigOverrideStr == "" && capabilitiesCfgOverride == nil {
+		return nil, nil
+	}
+	if capabilitiesCfgOverride == nil {
+		return &installConfigOverrideStr, nil
+	}
+	if installConfigOverrideStr == "" {
+		return capabilitiesCfgOverride, nil
+	}
+
+	installConfigOverrideStr, err = mergeJson(installConfigOverrideStr, *capabilitiesCfgOverride)
+	if err != nil {
+		return nil, err
 	}
+	return &installConfigOverrideStr, nil
+}
+
+// getInstallConfigOverrideForCapabilities generates install config override JSON for OpenShift capabilities configuration.
+// It automatically sets appropriate capabilities based on the platform type:
+// - Baremetal platforms: Sets baseline to "None" and includes default baremetal capabilities
+// - Non-baremetal platforms: Sets baseline to "vCurrent" and only includes user-specified capabilities
+// Returns nil if no capabilities configuration is needed (non-baremetal with empty capabilities).
+func getInstallConfigOverrideForCapabilities(oacp *controlplanev1alpha2.OpenshiftAssistedControlPlane, aci *hiveext.AgentClusterInstall) (*string, error) {
```

Comment:
Why make this a string pointer rather than string? I doesn't look like we have different behavior for empty vs nil so feels like a string would just be better.